File: T34-Tuplespaces-2025/Client-Python/command_processor.py
from typing import List
import re
class CommandProcessor:
    SPACE = " "
    BGN_TUPLE = "<"
    END_TUPLE = ">"
    PUT = "put"
    READ = "read"
    TAKE = "take"
    EXIT = "exit"
    GET_TUPLE_SPACES_STATE = "getTupleSpacesState"

    def __init__(self, client_service):
        self.client_service = client_service

    def parse_input(self):
        exit_flag = False
        while not exit_flag:
            try:
                line = input("> ").strip()
                split = line.split(self.SPACE)
                command = split[0]

                if command == self.PUT:
                    self.put(split)
                elif command == self.READ:
                    self.read(split)
                elif command == self.TAKE:
                    self.take(split)
                elif command == self.GET_TUPLE_SPACES_STATE:
                    self.get_tuple_spaces_state()
                elif command == self.EXIT:
                    exit_flag = True
                else:
                    self.print_usage()
            except EOFError:
                break

    def put(self, split: list[str]):
        # Check if the input is valid
        if not self.input_is_valid(split):
            self.print_usage()
            return

        tuple_data = split[1]
        delays = [0, 0, 0]

        # Handle optional delays
        if len(split) >= 3:
            valid = True
            for i in range(2, min(len(split), 5)):  # accepts split[2] to split[4]
                if split[i].isdigit():
                    delays[i - 2] = int(split[i])
                else:
                    valid = False
                    break
            if not valid:
                self.print_usage()
                return

        if self.client_service.put(tuple_data, delays):
            print("OK\n")
        else:
            self.print_usage()


    def read(self, split: list[str]):
        # Check if the input is valid
        if not self.input_is_valid(split):
            self.print_usage()
            return

        tuple_data = split[1]
        delays = [0, 0, 0]

        if len(split) >= 3:
            valid = True
            for i in range(2, min(len(split), 5)): # min garante que não exploda fora do seu range
                if split[i].isdigit():
                    delays[i - 2] = int(split[i])
                else:
                    valid = False
                    break
            if not valid:
                self.print_usage()
                return

        result = self.client_service.read(tuple_data, delays)
        if result:
            print(f"OK\n{result}\n")
        else:
            self.print_usage()


    def is_concrete_tuple(self, tuple_data: str) -> bool:
        return not any(substr in tuple_data for substr in ['[', ']', '.*', '|'])




    def take(self, split: list[str]):
        # Verifica se o input é válido
        if not self.input_is_valid(split):
            self.print_usage()
            return

        tuple_data = split[1]
        
        # Remove os < > 
        if not tuple_data.startswith("<") or not tuple_data.endswith(">"):
            self.print_usage()
            return

        pattern = tuple_data[1:-1]  # regex sem < >

        # Tenta compilar a regex para verificar se é válida
        try:
            re.compile(pattern)
        except re.error as e:
            print(f"Expressão regular inválida: {e}")
            return

        delays = [0, 0, 0]
        if len(split) >= 3:
            for i in range(2, min(len(split), 5)):
                if split[i].isdigit():
                    delays[i - 2] = int(split[i])
                else:
                    self.print_usage()
                    return

        # Chama o serviço e imprime o resultado
        result = self.client_service.take(tuple_data, delays)
        if result:
            print(f"OK\n{result}\n")


    def get_tuple_spaces_state(self):
        delays = [0, 0, 0]
        result = self.client_service.get_tuple_spaces_state(delays)

        if result is None:
            return

        print("OK")

        if not result:
            print("[]\n")
            return

        print("[", end="")
        for i in range(len(result) - 1):
            print(f"{result[i]}", end=", ")
        print(f"{result[-1]}]\n")


    def print_usage(self):
        print("Usage:\n"
              "- put <element[,more_elements]>\n"
              "- read <element[,more_elements]>\n"
              "- take <element[,more_elements]>\n"
              "- getTupleSpacesState\n"
              "- sleep <integer>\n"
              "- exit\n")

    def input_is_valid(self, input_data: List[str]):
        if (len(input_data) < 2
                or not input_data[1].startswith(self.BGN_TUPLE)
                or not input_data[1].endswith(self.END_TUPLE)):
            return False

        delay_count = len(input_data) - 2
        if delay_count < 0 or delay_count > 3:
            return False

        return True

File: T34-Tuplespaces-2025/Client-Java/src/main/java/pt/ulisboa/tecnico/tuplespaces/client/CommandProcessor.java
package pt.ulisboa.tecnico.tuplespaces.client;

import pt.ulisboa.tecnico.tuplespaces.client.grpc.ClientService;
import java.util.List;
import java.util.Scanner;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;


public class CommandProcessor {

    private static final String SPACE = " ";
    private static final String BGN_TUPLE = "<";
    private static final String END_TUPLE = ">";
    private static final String PUT = "put";
    private static final String READ = "read";
    private static final String TAKE = "take";
    private static final String SLEEP = "sleep";
    private static final String EXIT = "exit";
    private static final String GET_TUPLE_SPACES_STATE = "getTupleSpacesState";

    private final ClientService clientService;

    public CommandProcessor(ClientService clientService) {
        this.clientService = clientService;
    }

    void parseInput() {

        Scanner scanner = new Scanner(System.in);
        boolean exit = false;

        while (!exit) {
            System.out.print("> ");
            String line = scanner.nextLine().trim();
            String[] split = line.split(SPACE);
             switch (split[0]) {
                case PUT:
                    this.put(split);
                    break;

                case READ:
                    this.read(split);
                    break;

                case TAKE:
                    this.take(split);
                    break;

                case GET_TUPLE_SPACES_STATE:
                    this.getTupleSpacesState();
                    break;

                case SLEEP:
                    this.sleep(split);
                    break;

                case EXIT:
                    exit = true;
                    break;

                default:
                    this.printUsage();
                    break;
             }
        }
        scanner.close();
        clientService.shutdown();
    }

    private void put(String[] split) {
        if (!this.inputIsValid(split)) {
            this.printUsage();
            return;
        }
    
        String tuple = split[1];
        Integer[] delays = {0, 0, 0}; // standard delays
    
        if (split.length >= 3 && split.length <= 5) {
            boolean valid = true; 
            for (int i = 2; i < split.length; i++) {
                if (split[i].matches("\\d+")) { // só aceita números positivos
                    int delay = Integer.parseInt(split[i]);
                    delays[i - 2] = delay;
                } else {
                    valid = false;
                    break;
                }
            }
            if (!valid) {
                printUsage();
                return;
            }
        }
    
        boolean res = clientService.put(tuple, delays);
        if (!res) return;
        System.out.println("OK\n");
    }
    
    

    private void read(String[] split) {
        if (!inputIsValid(split)) {
            printUsage();
            return;
        }
    
        String pattern = split[1];
        Integer[] delays = {0, 0, 0};
    
        if (split.length >= 3) {
            for (int i = 2; i < split.length; i++) {
                if (!split[i].matches("\\d+")) {
                    printUsage();
                    return;
                }
                delays[i - 2] = Integer.parseInt(split[i]);
            }
        }
    
        String result = clientService.read(pattern, delays);
        if (result != null) {
            System.out.printf("OK%n%s%n", result);
        }
    }
    

    
    private void take(String[] split) {
        if (!inputIsValid(split)) {
            printUsage();
            return;
        }

        String tuple = split[1];

        // Verifica se a regex é válida (tenta compilar)
        String pattern = tuple.substring(1, tuple.length() - 1); // remove < >
        try {
            Pattern.compile(pattern);
        } catch (PatternSyntaxException e) {
            System.err.println("Expressão regular inválida: " + e.getMessage());
            return;
        }

        Integer[] delays = {0, 0, 0};

        if (split.length >= 3) {
            for (int i = 2; i < split.length; i++) {
                if (!split[i].matches("\\d+")) {
                    printUsage();
                    return;
                }
                delays[i - 2] = Integer.parseInt(split[i]);
            }
        }

        String result = clientService.take(tuple, delays);
        if (result != null) {
            System.out.printf("OK%n%s%n%n", result);
        }
    }


    private void getTupleSpacesState() {

        List<String> result = clientService.getTupleSpacesState();
        
        if (result == null) 
            return;

        System.out.println("OK");

        if (result.isEmpty()){
            System.out.println("[]\n");
            return;
        }
        int size = result.size() - 1;
        System.out.print("[");
        for (int i = 0; i < size; i++) System.out.print(result.get(i) + ", ");
        System.out.println(result.get(size++) + "]\n");
    }


    private void sleep(String[] split) {
      if (split.length != 2){
        this.printUsage();
        return;
      }
      Integer time;

      // checks if input String can be parsed as an Integer
      try {
         time = Integer.parseInt(split[1]);
      } catch (NumberFormatException e) {
        this.printUsage();
        return;
      }

      try {
        Thread.sleep(time*1000);
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
    }

    private void printUsage() {
        System.out.println("Usage:\n" +
                "- put <element[,more_elements]>\n" +
                "- read <element[,more_elements]>\n" +
                "- take <element[,more_elements]>\n" +
                "- getTupleSpacesState <server>\n" +
                "- sleep <integer>\n" +
                "- exit\n");
    }

    private boolean inputIsValid(String[] input) {
        if (input.length < 2
            || (!input[1].startsWith(BGN_TUPLE))
            || (!input[1].endsWith(END_TUPLE)))
            return false;
    
        int delayCount = input.length - 2;
        if (delayCount < 0 || delayCount > 3)
            return false;
    
        return true;
    }
}
